File: docs/article/source/scala/apps/cargo-tracker/app/cargotracker/tracking/domain/model/valueobjects/TrackingNumber.scala
package cargotracker.tracking.domain.model.valueobjects

/** 追跡番号（Tracking Context）。
  *
  *   - 形式: `TN-` + 6 桁ゼロ埋め整数（合計 9 文字）。data-model.md L787 の `VARCHAR(20)` 制約に整合
  *   - 採番ポリシー: ADR 0010 参照
  *   - 採番後の変更は不可（不変条件）
  */
opaque type TrackingNumber = String

object TrackingNumber:
  private val pattern = "^TN-\\d{6,18}$".r

  sealed trait Error
  case object InvalidFormat extends Error

  def apply(raw: String): Either[Error, TrackingNumber] =
    if pattern.matches(raw) then Right(raw)
    else Left(InvalidFormat)

  def unsafeFrom(raw: String): TrackingNumber =
    apply(raw).fold(_ => throw IllegalArgumentException(s"Invalid TrackingNumber: $raw"), identity)

  /** `BIGSERIAL` 値から `TN-000001` 形式の追跡番号を生成する（ADR 0010）。 */
  def fromSequence(seq: Long): TrackingNumber =
    f"TN-$seq%06d"

  extension (tn: TrackingNumber) def value: String = tn

File: apps/cargo-tracker/backend/trackingms/src/main/java/com/example/cargotracker/tracking/domain/model/valueobjects/TrackingNumber.java
package com.example.cargotracker.tracking.domain.model.valueobjects;

import com.example.cargotracker.shared.domain.error.BusinessRuleViolation;

/**
 * 追跡番号（trackingms の型）。
 *
 * <p><b>bookingms の {@code TrackingNumber} とは別の型にします</b>（domain-model.md
 * 「置かないもの」に識別子が挙がっている）。契約では文字列で運び、受け取った側が
 * 自分の型へ組み直します。共有すると、片方の BC が採番規則を変えたときにもう一方が
 * 巻き込まれます。</p>
 *
 * <p><b>採番しません。</b> 番号は bookingms の投影が採り、契約コマンドで届きます。</p>
 */
public record TrackingNumber(String value) {

    public TrackingNumber {
        if (value == null || value.isBlank()) {
            throw new BusinessRuleViolation("追跡番号は必須です");
        }
        value = value.trim();
    }

    public static TrackingNumber of(String value) {
        return new TrackingNumber(value);
    }
}
